graphql_compiler/cost_estimation/cardinality_estimator.py

```diff
@@ -0,0 +1,272 @@
+    return child_optional_depth > parent_optional_depth
+
+
+def _is_subexpansion_folded(location):
+    """Return True if location is the root of a folded subexpansion."""
+    return isinstance(location, FoldScopeLocation) and len(location.fold_path) == 1
+
+
+def _is_subexpansion_recursive(query_metadata, child_location, parent_location):
+    """Return True if child_location is the root of a recursive subexpansion."""
+    edge_direction, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
+    for recurse_info in query_metadata.get_recurse_infos(parent_location):
+        if recurse_info.edge_direction == edge_direction and recurse_info.edge_name == edge_name:
+            return True
+    return False
+
+
+def _get_all_original_child_locations(query_metadata, start_location):
+    """Get all original child Locations of a start Location and revisits to the start Location.
+
+    Args:
+        query_metadata: QueryMetadataTable object
+        start_location: Location object, where we're looking for children
+
+    Returns:
+        list of child Locations. Given start_location, get all revisits to start_location, then for
+        all visits, get all child locations and return ones that are original visits.
+    """
+    child_locations = set()
+
+    start_location_revisit_origin = [start_location]
+    start_location_revisits = list(query_metadata.get_all_revisits(start_location))
+
+    for location in chain(start_location_revisit_origin, start_location_revisits):
+        for child_location in query_metadata.get_child_locations(location):
+            child_location_revisit_origin = query_metadata.get_revisit_origin(child_location)
+            if child_location_revisit_origin is None:
+                # If child_location is not a revisit, set origin to child_location
+                child_location_revisit_origin = child_location
+            child_locations.add(child_location_revisit_origin)
+
+    return list(child_locations)
+
+
+def _get_last_edge_direction_and_name_to_location(location):
+    """Get the direction and name of the last edge to a non-root BaseLocation object."""
+    if isinstance(location, Location):
+        edge_direction, edge_name = get_edge_direction_and_name(location.query_path[-1])
+    elif isinstance(location, FoldScopeLocation):
+        edge_direction, edge_name = location.fold_path[-1]
+    else:
+        raise AssertionError(u'Unexpected location encountered: {}'.format(location))
+    return edge_direction, edge_name
+
+
+def _get_parent_and_child_name_from_edge(schema_graph, child_location):
+    """Get the base classname of a location and its parent from the last edge information."""
+    edge_direction, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
+    edge_element = schema_graph.get_edge_schema_element_or_raise(edge_name)
+    if edge_direction == EDGE_DESTINATION_PROPERTY_NAME:
```

Comment:
One nit: EDGE_DESTINATION_PROPERTY_NAME is an OrientDB variable that is very counterintuitive. (It is equal to 'out'). Could you instead use OUTBOUND_EDGE_DIRECTION and INBOUND_EDGE_DIRECTION? They are defined on graphql_compiler/compiler/helpers.py